### PR TITLE
Assignment6

### PR DESCRIPTION
**Pull request by FenrisStargazer**
abrand3 / FenrisStargazer, no partner
EC not attempted

### CODE REVIEW

**Review by Bhopss152 — COMMENTED** [review]
Create a Console project called "Calculate.". ✔
Define a Program Class✔
Define two init-only setter properties, WriteLine and ReadLine, that contain delegates for writing a line of text and reading a line of text respectively ✔
Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
Set the default behavior for the WriteLine and ReadLine properties to invoke System.Console versions of the methods and add an empty default constructor. ✔
Define a Calculator class ✔
Define static Add, Subtract, Multiple, and Divide methods that have two parameters and return a third parameter. ✔
Define a read-only property, MathematicalOperations, of type System.Collections.Generics.IReadOnlyDictionary<TKey,TValue> that:
is initialized to a System.Collections.Generics.Dictionary<<TKey,TValue> instance that. ✔
Uses char for the key corresponding to the operators +, -, *, and /. ✔
Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
Implement a TryCalculate method following "TryParse" pattern ✔
Valid calculation expressions include such strings as "3 + 4", "42 - 2", etc. ✔
If there is no whitespace around the operator, you can assume the calculation is invalid and return false. Similarly if the operands are not integers. ✔
Use string.Split(), pattern matching, logical and operators to parse the string in their entirety ✔
Index into the MathematicalOperations method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ✔
Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
Be sure to use the WriteLine/ReadLine properties on Program for testing the input and output of your program. ❌    

Very nice job on the code, I marked two things that you were missing which are a test and the writeLine/readLine properties. Fix those and itll be good to go!

**Comment by AndrewKorchemniy** [comment]
Will review

**Review by AndrewKorchemniy — COMMENTED** [review]
- Create a *Console* project called "Calculate.". ✔
- Define a Program Class
  - Define two init-only setter properties, `WriteLine` and `ReadLine`, that contain delegates for writing a line of text and reading a line of text respectively ✔
  - Write a test that sets these properties at construction time and then invokes the properties and verifies the expected behavior occurs. ✔
  - Set the default behavior for the `WriteLine` and `ReadLine` properties to invoke `System.Console` versions of the methods and add an empty default constructor. ✔
- Define a Calculator class ✔
  - Define static `Add`, `Subtract`, `Multiple`, and `Divide` methods that have two parameters and return a third parameter. ✔
  - Define a read-only property, `MathematicalOperations`, of type `System.Collections.Generics.IReadOnlyDictionary<TKey,TValue>` that:
    - is initialized to a `System.Collections.Generics.Dictionary<<TKey,TValue>` instance that. ✔
      - Uses `char` for the key corresponding to the operators +, -, *, and /. ✔
      - Has values that correspond with the Add, Subtract, Multiple, and Divide methods. ✔
  - Implement a `TryCalculate` method following "TryParse" pattern ✔
    - Valid `calculation` expressions include such strings as "3 + 4", "42 - 2", etc. ✔
    - If there is no whitespace around the operator, you can assume the `calculation` is invalid and return false. Similarly if the operands are not integers. ✔
    - Use `string.Split()`, pattern matching, logical and operators to parse the string in their entirety ❌
    - Index into the `MathematicalOperations` method using the operator parsed during pattern matching to find the corresponding implementation and invoke it. ❌
- Implement the Program class to instantiate the calculator and invoke it based on user input from the console. ✔
- Be sure to use the `WriteLine`/`ReadLine` properties on `Program` for testing the input and output of your program. ✔

- Place all shared project properties into a `Directory.Build.Props` file. ❌
- Place all shared project items into a `Directory.Build.targets` file. 
- nullable reference types is enabled  ❌
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers)  ❌
- Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional)   ❌
- and enabled .NET analyzers for both projects ❌
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ✔
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔

File: ProgramTests/CalculatorTests.cs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+        [TestMethod]
+        public void TestDivide()
+        {
+            double testDivideAct = Calculator.Divide(5, 2);
+            double testDivideExp = 2.5;
+            Assert.AreEqual<double>(testDivideExp, testDivideAct);
+        }
+
+        [TestMethod]
+        public void TestTryCalculateAddSuccess()
```
Comment (AndrewKorchemniy) [inline ProgramTests/CalculatorTests.cs:56]:
```suggestion
        public void TryCalculate_ValidAddExpression_Success()
```
Placing "Test" into the name of each test method isn't ideal.
There are different commonly used formats, but the following is the most clear to me:
[UnitOfWork_StateUnderTest_ExpectedBehavior]

Where UnitOfWork refers to the member you're testing. Which, in this case, is the TryCalculate() method.

File: Calculate/Program.cs [hunk 2]
```diff
@@ -0,0 +1,39 @@
+﻿namespace Calculate;
+public class Program
+{
+    public Action<string> WriteLine { get; init; }
+    public Func<string> ReadLine { get; init; }
+
+    public Program()
```
[the remark this answers is not in the document]
Comment (mmwoodfo) [reply Calculate/Program.cs:7]: This is the way it should be done, there should be a second constructor here for setting them. You should not rely on object initialization based on the instructions.

File: Calculate/Calculator.cs [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            {'*', Multiply},
+            {'-', Subtract},
+            {'+', Add}
+        };
+    }
+
+    public static double Add(int paramOne, int paramTwo)
+    {
+        return paramOne + paramTwo;
+    }
```
Comment (mmwoodfo) [inline Calculate/Calculator.cs:20]:
```suggestion
    public static double Add(int paramOne, int paramTwo) => paramOne + paramTwo;
```
You could do this in a single line, this is acceptable for simple methods that are just doing one thing.

File: Calculate/Calculator.cs [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+
+    public static double Divide(int paramOne, int paramTwo)
+    {
+        return (double) paramOne / paramTwo;
+    }
+
+    public static bool TryCalculate(string calculation, out double result)
+    {
+        bool TF = false;
```
Comment (mmwoodfo) [inline Calculate/Calculator.cs:39]:
```suggestion
        bool isValid = false;
```
Typically, you should name your bools with what they are expected to check. Some valid names would be `isValid` `calculationIsSuccessful` `isCalculated` etc.

File: Calculate/Calculator.cs [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        result = default;
+        Calculator calc = new();
+
+        string[] ints = calculation.Split(' ');
```
Comment (mmwoodfo) [inline Calculate/Calculator.cs:43]: You should be using the Split method along side pattern matching. e.g.
`if (expression?.Split(" ") is [string operator1, [char mathmaticalOperator], string operand2])`

**Pull request closed without merging**